[SQL] 'az sql db str-policy set/show': Add Set and Show ShortTermRetentionPolicy

lululilliancoding opened the pull request:
**Description<!--Mandatory-->**  
Add two new commands for getting and updating SQL DB short term retention. 

**Testing Guide**  
UPDATE command: 
az sql db str-policy set -g mygroup -s myserver -n mydb --retention-days retentionindays --diffbackupinterval-hours diffbackupintervalhours

GET command: 
az sql db str-policy show -g mygroup -s myserver -n mydb

**History Notes**  
[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.  
[Component Name 2] az command b: Add some customer-facing feature.

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

MyronFanQiu commented:
/azp run

yonzhan commented:
SQL

MyronFanQiu commented on src/azure-cli/azure/cli/command_modules/sql/commands.py:
```diff
@@ -28,6 +28,7 @@
     get_sql_database_blob_auditing_policies_operations,
     get_sql_database_long_term_retention_backups_operations,
     get_sql_database_long_term_retention_policies_operations,
+    get_sql_database_short_term_retention_policies_operations,
```
This function doesn't exist

lululilliancoding replied:
Thanks! Added.

lululilliancoding commented:
For the CLI Linter failure, should I fix both "FAIL - HIGH severity" and "FAIL - MEDIUM severity"? 
I found some of "FAIL - MEDIUM severity" are not related with my changes.

MyronFanQiu commented:
@lululilliancoding If you can solve it, better to do it. Severity medium won't block your PR. The below violation should be fixed.
![image](https://user-images.githubusercontent.com/49134743/91375051-399d7380-e84c-11ea-9f52-12e368e6c47e.png)

lululilliancoding commented:
@MyronFanQiu I got this error when I tried to use the parameter name "diff_backup_interval_in_hours" from SDK client locally: 
ValueError: The operation 'azure.mgmt.sql.operations#BackupShortTermRetentionPoliciesOperations.get_short_term_retention' is invalid.

Is this issue from SDK? Does this SDK released already? https://github.com/Azure/sdk-release-request/issues/638

lululilliancoding commented:
This scenario test failed in Checks which is not I edit: 
FAILED src/azure-cli/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py::SqlManagedInstanceDbShortTermRetentionScenarioTest::test_sql_managed_db_short_retention

lululilliancoding commented:
> @lululilliancoding If you can solve it, better to do it. Severity medium won't block your PR. The below violation should be fixed.
> ![image](https://user-images.githubusercontent.com/49134743/91375051-399d7380-e84c-11ea-9f52-12e368e6c47e.png)

MyronFanQiu commented:
https://github.com/Azure/azure-sdk-for-python/blob/release/v3/sdk/sql/azure-mgmt-sql/azure/mgmt/sql/operations/_managed_databases_operations.py#L40 You bump the dependency, which means you change all commands already. Since you're using a new sdk using a new api-version. You need to re-run these tests lively.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/sql/commands.py:
```diff
@@ -300,6 +301,18 @@ def load_command_table(self, _):
+                            client_factory=get_sql_backup_short_term_retention_policies_operations,
+                            is_preview=True) as g:
+
+        g.custom_command('set', 'update_short_term_retention', supports_no_wait=True)
```
I am thinking if we should use `update` as command name? May I know the operation is Patch or PUT?

lululilliancoding replied:
The swagger itself has both PUT and Patch

Juliehzl replied:
with support_no_wait, a wait command will be expected to add

Juliehzl commented on src/azure-cli/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py:
```diff
@@ -712,6 +712,46 @@ def test_sql_db_operation_mgmt(self, resource_group, resource_group_location, se
+    @ResourceGroupPreparer()
+    @SqlServerPreparer()
+    @AllowLargeResponse()
+    def test_sql_db_short_term_retention(self, resource_group, server):
```
you need to record for the test with yaml file provided. otherwise CI will fail.

lululilliancoding replied:
SqlServerDbShortTermRetentionScenarioTest doesn't pass yet locally. Should I still upload its yaml file? Looks like yaml file is auto generated after run the scenario test.

lululilliancoding replied:
> https://github.com/Azure/azure-sdk-for-python/blob/release/v3/sdk/sql/azure-mgmt-sql/azure/mgmt/sql/operations/_managed_databases_operations.py#L40 You bump the dependency, which means you change all commands already. Since you're using a new sdk using a new api-version. You need to re-run these tests lively.

I added new swagger shortTermRetentionPolicies.json to V2020-02-02-Preivew. 
The managedDatabases.json is not I added. But looks like since it is a new swagger in version V2020-02-02-Preivew so it was forced updated in new SDK. 

I might not familiar with other swagger's scenario tests. Is there any faster way to make it work?

lululilliancoding commented:
> https://github.com/Azure/azure-sdk-for-python/blob/release/v3/sdk/sql/azure-mgmt-sql/azure/mgmt/sql/operations/_managed_databases_operations.py#L40 You bump the dependency, which means you change all commands already. Since you're using a new sdk using a new api-version. You need to re-run these tests lively.

I added new swagger shortTermRetentionPolicies.json to V2020-02-02-Preivew.
The managedDatabases.json is not I added. But looks like since it is a new swagger in version V2020-02-02-Preivew so it was forced updated in new SDK.

I might not familiar with other swagger's scenario tests. Is there any faster way to make it work?

lululilliancoding commented:
Another question. 

Background: I added a new parameter 'diffbackupintervalinhours' for a new API but not deployed to Stage or any other clusters yet. 

Issue: When I run scenario test which I added, I got error : 
    _Query 'diffbackupintervalinhours' doesn't yield expected value '24', instead the actual value is '**None**'._

Question: Even I have sdk client support new parameter, but is this means I can't make the scenario test pass untill I have live API deployed on certain cluster?

Juliehzl commented on src/azure-cli/azure/cli/command_modules/sql/tests/latest/test_sql_commands.py:
```diff
@@ -729,6 +774,7 @@ def test_sql_db_long_term_retention(
         })
 
         # test update long term retention on live database
+        print('Creating MI...\n')
```
is it required?

lululilliancoding replied:
just fur debug. will remove it in next commit.

lululilliancoding replied:
removed

Juliehzl commented:
please rebase latest dev branch and the new SDK version is already here.

lululilliancoding commented:
The record is on Stage/Dodfood.

Juliehzl commented on src/azure-cli/azure/cli/command_modules/sql/custom.py:
```diff
@@ -2698,6 +2698,44 @@ def update_long_term_retention(
+    Updates short term retention for live database
+    '''
+
+    policy = client.create_or_update(
+        database_name=database_name,
+        server_name=server_name,
+        resource_group_name=resource_group_name,
+        retention_days=retention_days,
+        diff_backup_interval_in_hours=diffbackup_hours)
+
+    return policy
```
as you add support_no_wait flag for the command, you need to add sdk_no_wait() in custom function
```
# inside load_command_table(...)
with self.command_group('mymod', mymod_sdk) as g:
    g.custom_command('command2', 'do_something_2', supports_no_wait=True)

# inside custom.py
from azure.cli.core.util import sdk_no_wait
def do_something_2(cmd, arg1, arg2, no_wait=False):
    return sdk_no_wait(no_wait, client.create_or_update, arg1, arg2)
```

lululilliancoding replied:
changed, pl check.

Juliehzl commented:
@lululilliancoding LGTM now, but please make sure all commands work as expected and it would be better to add test command in your test scenarios including wait command/ --no-wait usage

lululilliancoding commented:
Set and Show command works as expected after this commit : https://github.com/Azure/azure-cli/pull/14919/commits/3b1906d6153d2f6d5cbbb5302991034f788e6a5a.
